### Цели и задачи проекта

Цель проекта - изучение развития игровой индустрии для написания статьи-исследования посредством ознакомления, предобработки и анализа данных за определенный период.

Основные задачи проекта:
- Загрузить и ознакомиться с данными.
- Исследовать названия столбцов датафрейма, типы данных, наличие пропусков, дубликаты.
- Отобрать данные по времени выхода игры, категоризовать игры по оценкам, выделить топ платформ по количеству игр.
- Сделать вывод.


### Описание данных
Данные `/datasets/new_games.csv` содержат информацию о продажах игр разных жанров и платформ, а также пользовательские и экспертные оценки игр:

- `Name` — название игры.
- `Platform` — название платформы.
- `Year of Release` — год выпуска игры.
- `Genre` — жанр игры.
- `NA sales` — продажи в Северной Америке (в миллионах проданных копий).
- `EU sales` — продажи в Европе (в миллионах проданных копий).
- `JP sales` — продажи в Японии (в миллионах проданных копий).
- `Other sales` — продажи в других странах (в миллионах проданных копий).
- `Critic Score` — оценка критиков (от 0 до 100).
- `User Score` — оценка пользователей (от 0 до 10).
- `Rating` — рейтинг организации ESRB (англ. Entertainment Software Rating Board). Эта ассоциация определяет рейтинг компьютерных игр и присваивает им подходящую возрастную категорию.

- **RP (Rating Pending)** — используется в рекламных материалах для игр, которым ещё не присвоен окончательный рейтинг ESRB. Принят в 1994 году и действует до сих пор.
- **EC (Early Childhood)**— игры, ориентированные на дошкольную аудиторию. Не содержат материалов, которые родители могли бы считать неприемлемыми. Принят в 1994 году и действовал до 2018 года.
- **E (Everyone)** — игры для всех возрастов. Могут содержать незначительное количество насилия и сленга. До 1998 года рейтинг был известен как Kids to Adults (K-A). Принят в 1994 году и действует до сих пор.
- **E10+** **(Everyone 10 and older)** — игры для людей от 10 лет и старше. Могут содержать более высокий уровень насилия, грубую разговорную речь. Принят в 2005 году и действует до сих пор.
- **T (Teen)** — игры для людей от 13 лет и старше. Могут содержать умеренное количество насилия, включая небольшие количества крови, грубую разговорную речь и юмор, сексуальное содержание. Принят в 1994 году и действует до сих пор.
- **M (Mature)** — игры для лиц от 17 лет и старше. Могут содержать более интенсивные или реалистичные изображения насилия, включая кровь, ужасы, увечья и изображения смерти, непристойные сексуальные сцены, частичную наготу и грубую ненормативную лексику. Принят в 1994 году и действует до сих пор.
- **AO (Adults Only)**— игры для людей от 18 лет и старше. Содержат материалы с более высоким уровнем воздействия, чем рейтинг M. Принят в 1994 году и действует до сих пор.

### Содержимое проекта

1. Загрузка данных и знакомство с ними.
2. Проверка ошибок и предобработка данных.
3. Фильтрация данных.
4. Категоризация данных.
5. Анализ данных и выводы.
---

## 1. Загрузка данных и знакомство с ними

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('/datasets/new_games.csv')

In [4]:
df.head(5)

,Name,Platform,Year of Release,Genre,NA sales,EU sales,JP sales,Other sales,Critic Score,User Score,Rating
0,Wii Sports,Wii,2006.0,Sports,41.36,28.96,3.77,8.45,76.0,8,E
1,Super Mario Bros.,NES,1985.0,Platform,29.08,3.58,6.81,0.77,NaN,NaN,NaN
2,Mario Kart Wii,Wii,2008.0,Racing,15.68,12.76,3.79,3.29,82.0,8.3,E
3,Wii Sports Resort,Wii,2009.0,Sports,15.61,10.93,3.28,2.95,80.0,8,E
4,Pokemon Red/Pokemon Blue,GB,1996.0,Role-Playing,11.27,8.89,10.22,1.00,NaN,NaN,NaN


In [5]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16954 non-null  object 
 1   Platform         16956 non-null  object 
 2   Year of Release  16681 non-null  float64
 3   Genre            16954 non-null  object 
 4   NA sales         16956 non-null  float64
 5   EU sales         16956 non-null  object 
 6   JP sales         16956 non-null  object 
 7   Other sales      16956 non-null  float64
 8   Critic Score     8242 non-null   float64
 9   User Score       10152 non-null  object 
 10  Rating           10085 non-null  object 
dtypes: float64(4), object(7)
memory usage: 1.4+ MB
None


Как видим, датафрейм содержит 16956 записей, в нем 11 столбцов. Памяти занял примерно 1.4 МБ. 4 столбца содержат данные типа float64, а остальные 7 содержат данные типа object (скорее всего в них содержится текстовая информация). Названия столбцов соответствуют описанию данных. Пропуски встречаются (в оценках и рейтинге точно). Странно, что для года выпуска используется float64, а не int64. EU sales, JP sales почему-то являются объектами, хотя должны быть float64 как NA sales. С User Score та же история.

Названия столбцов вполне сносные по смыслу, но для работы стоит привести к виду Snake Case. По поводу пропусков: очень много незаполненных оценок и рейтингов, иногда не хватает года релиза, названия, жанра. Данные о продажах есть в каждой строке.

---

## 2.  Проверка ошибок в данных и их предобработка


### 2.1. Названия, или метки, столбцов датафрейма

In [6]:
print(df.columns)

Index(['Name', 'Platform', 'Year of Release', 'Genre', 'NA sales', 'EU sales',
       'JP sales', 'Other sales', 'Critic Score', 'User Score', 'Rating'],
      dtype='object')


In [7]:
df.columns = df.columns.str.lower().str.replace(' ', '_')

In [8]:
print(df.columns)

Index(['name', 'platform', 'year_of_release', 'genre', 'na_sales', 'eu_sales',
       'jp_sales', 'other_sales', 'critic_score', 'user_score', 'rating'],
      dtype='object')


In [9]:
tmp = df.copy() 
len(tmp)

16956

### 2.2. Типы данных

Некорректные типы данных вероятно возникли из-за не слишком аккуратного импорта их из каких-либо баз данных или систем. После занесения в датафрейм типы данных не всегда меняли на удобный.

In [10]:
print(df['eu_sales'].unique())
print(df['jp_sales'].unique())
print(df['user_score'].unique())

['28.96' '3.58' '12.76' '10.93' '8.89' '2.26' '9.14' '9.18' '6.94' '0.63'
 '10.95' '7.47' '6.18' '8.03' '4.89' '8.49' '9.09' '0.4' '3.75' '9.2'
 '4.46' '2.71' '3.44' '5.14' '5.49' '3.9' '5.35' '3.17' '5.09' '4.24'
 '5.04' '5.86' '3.68' '4.19' '5.73' '3.59' '4.51' '2.55' '4.02' '4.37'
 '6.31' '3.45' '2.81' '2.85' '3.49' '0.01' '3.35' '2.04' '3.07' '3.87'
 '3.0' '4.82' '3.64' '2.15' '3.69' '2.65' '2.56' '3.11' '3.14' '1.94'
 '1.95' '2.47' '2.28' '3.42' '3.63' '2.36' '1.71' '1.85' '2.79' '1.24'
 '6.12' '1.53' '3.47' '2.24' '5.01' '2.01' '1.72' '2.07' '6.42' '3.86'
 '0.45' '3.48' '1.89' '5.75' '2.17' '1.37' '2.35' '1.18' '2.11' '1.88'
 '2.83' '2.99' '2.89' '3.27' '2.22' '2.14' '1.45' '1.75' '1.04' '1.77'
 '3.02' '2.75' '2.16' '1.9' '2.59' '2.2' '4.3' '0.93' '2.53' '2.52' '1.79'
 '1.3' '2.6' '1.58' '1.2' '1.56' '1.34' '1.26' '0.83' '6.21' '2.8' '1.59'
 '1.73' '4.33' '1.83' '0.0' '2.18' '1.98' '1.47' '0.67' '1.55' '1.91'
 '0.69' '0.6' '1.93' '1.64' '0.55' '2.19' '1.11' '2.29' '2.5' '0.96'
 '

В информации о продажах есть строки со значение 'unknown', т.е. неизвесто. Не в каждом регионе удалось собрать данные вероятно. В оценках от пользователей есть значение nan - not a number, пропуск, и 'tbd' - по идее to be determined, т.е. игра еще не получила достаточное количество отзывов от пользователей. Некоторые игры еще не получили оценки или получили их слишком мало.

In [11]:
for col in ['na_sales', 'eu_sales', 'jp_sales', 'other_sales', 'critic_score', 'user_score']:
    df[col] = df[col].astype(str).str.replace(',', '.').str.strip()
    df[col] = pd.to_numeric(df[col], errors='coerce')

In [12]:
df['year_of_release'] = pd.to_numeric(df['year_of_release'], errors='coerce').astype('Int64')

In [13]:
print(df.dtypes)

name                object
platform            object
year_of_release      Int64
genre               object
na_sales           float64
eu_sales           float64
jp_sales           float64
other_sales        float64
critic_score       float64
user_score         float64
rating              object
dtype: object


In [14]:
print(df.head(5))

                       name platform  year_of_release         genre  na_sales  \
0                Wii Sports      Wii             2006        Sports     41.36   
1         Super Mario Bros.      NES             1985      Platform     29.08   
2            Mario Kart Wii      Wii             2008        Racing     15.68   
3         Wii Sports Resort      Wii             2009        Sports     15.61   
4  Pokemon Red/Pokemon Blue       GB             1996  Role-Playing     11.27   

   eu_sales  jp_sales  other_sales  critic_score  user_score rating  
0     28.96      3.77         8.45          76.0         8.0      E  
1      3.58      6.81         0.77           NaN         NaN    NaN  
2     12.76      3.79         3.29          82.0         8.3      E  
3     10.93      3.28         2.95          80.0         8.0      E  
4      8.89     10.22         1.00           NaN         NaN    NaN  


### 2.3. Наличие пропусков в данных

In [15]:
print(df.isnull().sum())
print(sum(df.isnull().sum()))


name                  2
platform              0
year_of_release     275
genre                 2
na_sales              0
eu_sales              6
jp_sales              4
other_sales           0
critic_score       8714
user_score         9268
rating             6871
dtype: int64
25142


In [16]:
print(df.isnull().mean())

name               0.000118
platform           0.000000
year_of_release    0.016218
genre              0.000118
na_sales           0.000000
eu_sales           0.000354
jp_sales           0.000236
other_sales        0.000000
critic_score       0.513918
user_score         0.546591
rating             0.405225
dtype: float64


In [17]:
def show_missing_stats(tmp0):
    """
    Функция для отображения статистики пропущенных значений в DataFrame.
    """
    missing_stats = pd.DataFrame({
        'Кол-во пропусков': tmp0.isnull().sum(),
        'Доля пропусков': tmp0.isnull().mean()
    })
    missing_stats = missing_stats[missing_stats['Кол-во пропусков'] > 0]
    
    if missing_stats.empty:
        return "Пропусков в данных нет"
    
    return (missing_stats.style.format({'Доля пропусков': '{:.4f}'}).background_gradient(cmap='coolwarm'))
show_missing_stats(df)

,Кол-во пропусков,Доля пропусков
name,2,0.0001
year_of_release,275,0.0162
genre,2,0.0001
eu_sales,6,0.0004
jp_sales,4,0.0002
critic_score,8714,0.5139
user_score,9268,0.5466
rating,6871,0.4052


In [18]:
print(df.isnull().mean() * 100)

name                0.011795
platform            0.000000
year_of_release     1.621845
genre               0.011795
na_sales            0.000000
eu_sales            0.035386
jp_sales            0.023590
other_sales         0.000000
critic_score       51.391838
user_score         54.659118
rating             40.522529
dtype: float64


40-54% пропусков у оценок с рейтингами - вероятно проблемы со сбором данных, либо слишком много различных площадок, где у игры могут быть абсолютно разные оценки. В остальных столбцах максимум 1.6% пропусков. Строки с пропущенными именем, жанром, годом релиза стоит удалить - это важные параметры, которые определяют игру. С пропущенной информацией о продажах заменить на среднее значение в зависимости от названия платформы, жанра и года выхода игры, ведь можно представить насколько была популярна такая игра в определенное время. Пропущенные оценки и рейтинг скорее всего заменить на индикатор. 

In [19]:
df_clean = df.dropna(subset=['name', 'genre', 'year_of_release']).copy()

print(f"Удалено строк с пропущенными name/genre/year: {len(df) - len(df_clean)}")
print(f"Осталось строк: {len(df_clean)}")

Удалено строк с пропущенными name/genre/year: 277
Осталось строк: 16679


In [21]:
sales_cols = ['eu_sales', 'jp_sales']
for col in sales_cols:
    avg_sales = df_clean.groupby(['platform', 'year_of_release'])[col].transform('mean')
    df_clean[col] = df_clean[col].fillna(avg_sales)

In [22]:
df_clean['critic_score'] = df_clean['critic_score'].fillna(-1)
df_clean['user_score'] = df_clean['user_score'].fillna(-1)
df_clean['rating'] = df_clean['rating'].fillna('unknown')

In [23]:
print(df_clean.isnull().sum())
print(f"Осталось строк: {len(df_clean)}")

name               0
platform           0
year_of_release    0
genre              0
na_sales           0
eu_sales           0
jp_sales           0
other_sales        0
critic_score       0
user_score         0
rating             0
dtype: int64
Осталось строк: 16679


### 2.4. Явные и неявные дубликаты в данных

In [24]:
categorical_cols = ['name', 'platform', 'genre', 'rating']
for col in categorical_cols:
    print(f"\n{col}:")
    print(f"  Уникальных значений: {df_clean[col].nunique()}")
    print(f"  Первые 35 уникальных значений: {df_clean[col].unique()[:35]}")

print('\n')
platforms_lower = df_clean['platform'].str.lower().unique()
print(f"  После приведения к нижнему регистру platform: {len(platforms_lower)} уникальных значений")
print(f"  Значения: {sorted(platforms_lower)}")
print('\n')
genres_lower = df_clean['genre'].str.lower().str.strip().unique()
print(f"  После приведения к нижнему регистру и удаления пробелов genre: {len(genres_lower)} уникальных значений")
print(f"  Значения: {sorted(genres_lower)}")
print('\n')
print("\nРейтинги:")
ratings = df_clean['rating'].replace('K-A', 'E').unique()
print(f"  После замены: {len(ratings)} уникальных значений")
print(f"  Значения: {sorted(ratings)}")




name:
  Уникальных значений: 11426
  Первые 35 уникальных значений: ['Wii Sports' 'Super Mario Bros.' 'Mario Kart Wii' 'Wii Sports Resort'
 'Pokemon Red/Pokemon Blue' 'Tetris' 'New Super Mario Bros.' 'Wii Play'
 'New Super Mario Bros. Wii' 'Duck Hunt' 'Nintendogs' 'Mario Kart DS'
 'Pokemon Gold/Pokemon Silver' 'Wii Fit' 'Kinect Adventures!'
 'Wii Fit Plus' 'Grand Theft Auto V' 'Grand Theft Auto: San Andreas'
 'Super Mario World' 'Brain Age: Train Your Brain in Minutes a Day'
 'Pokemon Diamond/Pokemon Pearl' 'Super Mario Land' 'Super Mario Bros. 3'
 'Grand Theft Auto: Vice City' 'Pokemon Ruby/Pokemon Sapphire'
 'Brain Age 2: More Training in Minutes a Day'
 'Pokemon Black/Pokemon White' 'Gran Turismo 3: A-Spec'
 'Call of Duty: Modern Warfare 3'
 'Pokémon Yellow: Special Pikachu Edition' 'Call of Duty: Black Ops 3'
 'Call of Duty: Black Ops' 'Pokemon X/Pokemon Y'
 'Call of Duty: Black Ops II' 'Call of Duty: Modern Warfare 2']

platform:
  Уникальных значений: 31
  Первые 35 уникальных з

In [25]:
df_clean['name'] = df_clean['name'].str.lower().str.strip()
df_clean['platform'] = df_clean['platform'].str.lower().str.strip()
df_clean['genre'] = df_clean['genre'].str.lower().str.strip()
df_clean['rating'] = df_clean['rating'].replace('K-A', 'E')

In [26]:
duplicates_full = df_clean.duplicated()
print(f"\nПолных дубликатов строк (все столбцы): {duplicates_full.sum()}")

duplicates_by_name = df_clean.duplicated(subset=['name'], keep=False)
print(f"\nДубликатов по названию игры: {duplicates_by_name.sum()}")

duplicated_names = df_clean[df_clean.duplicated(subset=['name'], keep=False)]
if len(duplicated_names) > 0:
    print("\nПримеры игр с одинаковыми названиями:")
    print(duplicated_names[['name', 'platform', 'year_of_release', 'genre']].head(10))
    


Полных дубликатов строк (все столбцы): 235

Дубликатов по названию игры: 8129

Примеры игр с одинаковыми названиями:
                              name platform  year_of_release     genre
1                super mario bros.      nes             1985  platform
5                           tetris       gb             1989    puzzle
16              grand theft auto v      ps3             2013    action
17   grand theft auto: san andreas      ps2             2004    action
18               super mario world     snes             1990  platform
22             super mario bros. 3      nes             1988  platform
23              grand theft auto v     x360             2013    action
24     grand theft auto: vice city      ps2             2002    action
29  call of duty: modern warfare 3     x360             2011   shooter
31       call of duty: black ops 3      ps4             2015   shooter


In [27]:
duplicates_key = df_clean.duplicated(subset=['name', 'platform', 'year_of_release'], keep=False)
print(f"\nДубликатов по комбинации (название + платформа + год): {duplicates_key.sum()}")

if duplicates_full.sum() > 0:
    print(f"\nПолные дубликаты строк:")
    print(df_clean[duplicates_full])
    
    df_clean = df_clean.drop_duplicates()
    print(f"\nПолные дубликаты удалены. Осталось строк: {len(df_clean)}")
else:
    print("\nПолных дубликатов строк не найдено")


Дубликатов по комбинации (название + платформа + год): 472

Полные дубликаты строк:
                                                name platform  \
268                            batman: arkham asylum      ps3   
368                 james bond 007: agent under fire      ps2   
717                            god of war: ascension      ps3   
823                                wipeout: the game      wii   
848                  rayman raving rabbids: tv party      wii   
...                                              ...      ...   
16671        fullmetal alchemist: prince of the dawn      wii   
16753                                      routes pe      ps2   
16799                            transformers: prime      wii   
16912  metal gear solid v: the definitive experience     xone   
16940                          the longest 5 minutes      psv   

       year_of_release      genre  na_sales  eu_sales  jp_sales  other_sales  \
268               2009     action      2.24      1.31 

In [28]:
name_platform_counts = df_clean.groupby('name')['platform'].nunique()
multi_platform_games = name_platform_counts[name_platform_counts > 1]
print(f"\nИгр, выпущенных более чем на одной платформе: {len(multi_platform_games)}")

if len(multi_platform_games) > 0:
    print("\nПримеры популярных игр на разных платформах:")
    popular_multi = multi_platform_games.head(10)
    for game in popular_multi.index:
        platforms = df_clean[df_clean['name'] == game]['platform'].unique()
        years = df_clean[df_clean['name'] == game]['year_of_release'].unique()
        print(f"  '{game}': платформы {sorted(platforms)}, годы {sorted(years)}")


Игр, выпущенных более чем на одной платформе: 2748

Примеры популярных игр на разных платформах:
  '007: quantum of solace': платформы ['ds', 'pc', 'ps2', 'ps3', 'wii', 'x360'], годы [2008]
  '007: the world is not enough': платформы ['n64', 'ps'], годы [2000]
  '11eyes: crossover': платформы ['psp', 'x360'], годы [2009, 2010]
  '18 wheeler: american pro trucker': платформы ['gc', 'ps2'], годы [2001, 2002]
  '187: ride or die': платформы ['ps2', 'xb'], годы [2005]
  '2 in 1 combo pack: sonic heroes / super monkey ball deluxe': платформы ['x360', 'xb'], годы [2004, 2013]
  '2002 fifa world cup': платформы ['gc', 'ps2', 'xb'], годы [2002]
  '2010 fifa world cup south africa': платформы ['ps3', 'psp', 'wii', 'x360'], годы [2010]
  '2014 fifa world cup brazil': платформы ['ps3', 'x360'], годы [2014]
  '25 to life': платформы ['ps2', 'xb'], годы [2006]


In [29]:
key_cols = ['name', 'platform', 'year_of_release']
duplicate_groups = df_clean[df_clean.duplicated(subset=key_cols, keep=False)]

if len(duplicate_groups) > 0:
    print(f"\nНайдены группы записей с одинаковыми {key_cols}:")
    for _, group in duplicate_groups.groupby(key_cols):
        if len(group) > 1:
            print(f"\n  Группа: {group.iloc[0]['name']} - {group.iloc[0]['platform']} - {group.iloc[0]['year_of_release']}")
            print(f"    Количество записей: {len(group)}")
            print(f"    Продажи (NA, EU, JP, Other):")
            for _, row in group.iterrows():
                print(f"      {row['na_sales']}, {row['eu_sales']}, {row['jp_sales']}, {row['other_sales']}")
    
    print("\nОставляем запись с максимальными общими продажами для каждой группы")
    
    df_clean['total_sales'] = df_clean['na_sales'] + df_clean['eu_sales'] + df_clean['jp_sales'] + df_clean['other_sales']
    df_clean = df_clean.sort_values('total_sales', ascending=False).drop_duplicates(subset=key_cols, keep='first')
    df_clean = df_clean.drop('total_sales', axis=1)
    
    print(f"После обработки осталось строк: {len(df_clean)}")
else:
    print("\nАномальных дубликатов не найдено")



Найдены группы записей с одинаковыми ['name', 'platform', 'year_of_release']:

  Группа: madden nfl 13 - ps3 - 2012
    Количество записей: 2
    Продажи (NA, EU, JP, Other):
      2.11, 0.22, 0.0, 0.23
      0.0, 0.01, 0.0, 0.0

Оставляем запись с максимальными общими продажами для каждой группы
После обработки осталось строк: 16443


- Удалено строк с пропущенными name/genre/year: 277
- Удалено дубликатов строк (все столбцы): 235
- Удалено записей с одинаковыми ['name', 'platform', 'year_of_release']: 1 (оставили с максимальными общими продажами)

- В процессе подготовки данных вы могли что-либо удалять, например строки с пропусками или ошибками, дубликаты и прочее. В этом случае посчитайте количество удалённых строк в абсолютном и относительном значениях.

In [30]:
df_initial = pd.read_csv('/datasets/new_games.csv')

df_initial.columns = df_initial.columns.str.lower().str.replace(' ', '_')

initial_count = len(df_initial)
print(f"Начальное количество строк: {initial_count}")

dropped = initial_count - len(df_clean)
pct_dropped = (dropped / initial_count) * 100

print(f"  Удалено строк: {dropped}")
print(f"  Процент от исходных: {pct_dropped:.2f}%")
print(f"  Осталось строк: {len(df_clean)}")

Начальное количество строк: 16956
  Удалено строк: 513
  Процент от исходных: 3.03%
  Осталось строк: 16443


In [31]:
a, b = len(tmp), len(df_clean)
print(" Было строк в исходном датасете", a,
      '\n', "Осталось строк в датасете после обработки", b,
      '\n', "Удалено строк в датасете после обработки", a-b,
      '\n', "Процент потерь", round((a-b)/a*100, 2))

 Было строк в исходном датасете 16956 
 Осталось строк в датасете после обработки 16443 
 Удалено строк в датасете после обработки 513 
 Процент потерь 3.03


В итоге удалили строки без важной информации, избавились от явных и неявных дубликатов. Привели записи к хорошему виду. Всего потеряли 3% записей, что не так критично. Поэтому предобработанные данные можно использовать для фильтрации и категоризации.

1. Названия столбцов:
- Все столбцы привели к виду snake_case.
- Переименовали 11 столбцов (Year of Release - year_of_release, NA sales - na_sales и т.п.).

2. Типы данных:
- Year of Release, JP sales, EU sales, User Score имели не совсем корректные типы.
- Year of Release переведен в int64 для удобства, т.к. год целочисленный. Остальные приведены к виду float64 для обеспечения возможности обрабатывать и анализировать эти данные.
- Строковые значения 'unknown' и 'tbd' были найдены, заменены на Not a Number.

3. Пропуски:
- Пропуски скорее всего возникли из-за отсутствия тех или иных данных для игры, или же из-за неудачного сбора данных (технические проблемы или ошибки).
- Удалены строки с пропусками в name, genre, year - критически важные параметры; eu_sales и jp_sales заполнены средним по группам платформы и года - можно примерно предположить насколько были популярны игры для платформ в определенный момент; user_score и critic_score заполнены конкретным значением "-1", а rating заполнен 'unknown' - индикаторы-заглушки, не мешающие работать с данными.

4. Дубликаты:
- Неявные дубликаты найдены в жанрах (некоторые прописными, некоторые строчными написаны) и рейтингах (присутствовал устаревший рейтинг, который дублирует другой новый). В рейтинге было 9 уникальных значений, стало 8. В жанрах было 24 уникальных значения, стало 12.
- Найдено и удалено полностью явных 235 дубликатов и 1 запись дублирующая по названию, жанру, платформе и году.

5. Итого:
- Было строк: 16956
- Удалено: 513 строк (3.03%)
- Осталось строк: 16443

---

## 3. Фильтрация данных

In [32]:
df_actual = df_clean[(df_clean['year_of_release'] >= 2000) & (df_clean['year_of_release'] <= 2013)].copy()

In [33]:
year_counts = df_actual['year_of_release'].value_counts().sort_index()
for year, count in year_counts.items():
    print(f"  {int(year)}: {count} игр ({count/len(df_actual)*100:.2f}%)")

  2000: 350 игр (2.74%)
  2001: 482 игр (3.77%)
  2002: 829 игр (6.49%)
  2003: 775 игр (6.06%)
  2004: 762 игр (5.96%)
  2005: 939 игр (7.35%)
  2006: 1006 игр (7.87%)
  2007: 1197 игр (9.37%)
  2008: 1427 игр (11.17%)
  2009: 1426 игр (11.16%)
  2010: 1255 игр (9.82%)
  2011: 1136 игр (8.89%)
  2012: 652 игр (5.10%)
  2013: 544 игр (4.26%)


In [34]:
print(df_actual.head(10))
print('\n')
print(len(df_actual))

                         name platform  year_of_release       genre  na_sales  \
0                  wii sports      wii             2006      sports     41.36   
2              mario kart wii      wii             2008      racing     15.68   
3           wii sports resort      wii             2009      sports     15.61   
6       new super mario bros.       ds             2006    platform     11.28   
7                    wii play      wii             2006        misc     13.96   
8   new super mario bros. wii      wii             2009    platform     14.44   
10                 nintendogs       ds             2005  simulation      9.05   
11              mario kart ds       ds             2005      racing      9.71   
13                    wii fit      wii             2007      sports      8.92   
14         kinect adventures!     x360             2010        misc     15.00   

    eu_sales  jp_sales  other_sales  critic_score  user_score   rating  
0      28.96      3.77         8.45

---

## 4. Категоризация данных

In [35]:
def categorize_user_score(score):
    if score == -1:
        return 'неизвестно'
    elif 8 <= score <= 10:
        return 'высокая оценка'
    elif 3 <= score < 8:
        return 'средняя оценка'
    elif 0 <= score < 3:
        return 'низкая оценка'
    else:
        return 'неизвестно'

df_actual['user_score_category'] = df_actual['user_score'].apply(categorize_user_score)

print(df_actual['user_score_category'].value_counts())

неизвестно        6298
средняя оценка    4080
высокая оценка    2286
низкая оценка      116
Name: user_score_category, dtype: int64


In [36]:
def categorize_critic_score(score):
    if score == -1:
        return 'неизвестно'
    elif 80 <= score <= 100:
        return 'высокая оценка'
    elif 30 <= score < 80:
        return 'средняя оценка'
    elif 0 <= score < 30:
        return 'низкая оценка'
    else:
        return 'неизвестно'

df_actual['critic_score_category'] = df_actual['critic_score'].apply(categorize_critic_score)

print(df_actual['critic_score_category'].value_counts())

неизвестно        5612
средняя оценка    5422
высокая оценка    1691
низкая оценка       55
Name: critic_score_category, dtype: int64


In [37]:
top_platforms = df_actual['platform'].value_counts().head(7)

print("Топ-7 платформ по количеству игр:")
print(top_platforms)
print("\nВ процентном соотношении:")
print((top_platforms / len(df_actual) * 100).round(2))

Топ-7 платформ по количеству игр:
ps2     2127
ds      2120
wii     1275
psp     1180
x360    1121
ps3     1086
gba      811
Name: platform, dtype: int64

В процентном соотношении:
ps2     16.64
ds      16.59
wii      9.98
psp      9.23
x360     8.77
ps3      8.50
gba      6.35
Name: platform, dtype: float64


---

## 5. Итоговый вывод

1. Цель проекта:
Изучение развития игровой индустрии для написания статьи-исследования посредством ознакомления, предобработки и анализа данных за определенный период.

2. Общая характеристика данных:
- Исходный датасет из 11 столбцов с 16956 строками.
- Датасет соответствует описанию, но в записях встречаются пропуски или ошибки.
- Пропуски, дубликаты и ошибки присутствуют. Записи с критичными пропусками удалены, отсутствующие продажи заменены средним по группам, в оценках и рейтинге поставлены индикаторы-заглушки. Явные и неявные дубликаты обработаны. Явные удалены, неявные приведены к общему виду. Ошибки исправлены.

3. Детализация проведенных работ:
- Преобразованы типы данных: Год релиза с float64 на int64, JP sales, EU sales и User Score с object на float64. Замена пропусков: пропуски в EU sales и JP sales заменены на среднее по группам, в оценках и рейтинге на индикаторы-заглушки ("-1" и "unknown"), записи с пропусками в Name, Year of Release, Platform, Genre удалены. Устранение дубликатов: явные дубликаты удалены, неявные приведены к общему виду. Текстовые данные, кроме рейтинга, нормализованы, приведены к виду нижнего регистра, пробелы удалены (не у названий).
- Кол-во обнаруженных и исправленных пропусков: 25142 (277 записей с крит. пропусками удалены); Кол-во дубликатов: 236 записей удалено + 13 уникальных значений исправлено; Название одного рейтинга исправлено.

4. Описание нового срезанного датасета:
- Отобраны данные за 2000-2013 года включительно.
- В итоге отобрано 12780 записей.
  2000: 350 игр (2.74%),
  2001: 482 игр (3.77%),
  2002: 829 игр (6.49%),
  2003: 775 игр (6.06%),
  2004: 762 игр (5.96%),
  2005: 939 игр (7.35%),
  2006: 1006 игр (7.87%),
  2007: 1197 игр (9.37%),
  2008: 1427 игр (11.17%),
  2009: 1426 игр (11.16%),
  2010: 1255 игр (9.82%),
  2011: 1136 игр (8.89%),
  2012: 652 игр (5.10%),
  2013: 544 игр (4.26%).

5. Категоризация данных:
- Данные категоризованы по оценкам пользователей так: высокая оценка (от 8 до 10 включительно), средняя оценка (от 3 до 8, не включая правую границу интервала) и низкая оценка (от 0 до 3, не включая правую границу интервала). По оценкам критиков так: высокая оценка (от 80 до 100 включительно), средняя оценка (от 30 до 80, не включая правую границу интервала) и низкая оценка (от 0 до 30, не включая правую границу интервала).
- Получился следующий топ-7 платформ: ps2 - 2127 игр, ds - 2120 игр, wii - 1275 игр, psp - 1180 игр, x360 - 1121 игр, ps3 - 1086 игр, gba - 811 игр.

6. Заключительное резюме:
- Очищенные и скорректированные данные готовые для дальнейшей аналитики.
- Важно держать в уме, что в некоторых столбцах все так же присутствуют пропуски и заглушки, некоторые данные все равно неизвестны из-за качества исходного датасета.